File: LeetCode/0049_Group_Anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs) == 1:
            return [strs]
        
        anagram_groups = {}
        
        for word in strs:
            sorted_word = "".join(sorted(word))
            
            if (sorted_word in anagram_groups):
                anagram_groups[sorted_word].append(word)
            else:
                anagram_groups[sorted_word] = [word]
                
        return anagram_groups.values()

File: LeetCode/Java/0049-Group-Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0)
            return new ArrayList();
            
        HashMap<String, List> anagramGroups = new HashMap<>();

        for (String word: strs) {
            char[] wordArray = word.toCharArray();
            Arrays.sort(wordArray);
            String sortedWord = String.valueOf(wordArray);
            
            if (!anagramGroups.containsKey(sortedWord))
                anagramGroups.put(sortedWord, new ArrayList());
            
            anagramGroups.get(sortedWord).add(word);
        }
        
        return new ArrayList(anagramGroups.values());
    }
}
